toonz/sources/include/tools/track.h

```diff
@@ -0,0 +1,409 @@
+        { b = c; bb = cc; } else { a = c; aa = cc; }
+    }
+    return bb - aa >= 0.5*TConsts::epsilon ? (double)a + (value - aa)/(bb - aa) : (double)a;
+  }
+
+public:
+  inline double indexByOriginalIndex(double originalIndex) const
+    { return binarySearch<&TTrackPoint::originalIndex>(originalIndex); }
+  inline double indexByTime(double time) const
+    { return binarySearch<&TTrackPoint::time>(time); }
+  inline double indexByLength(double length) const
+    { return binarySearch<&TTrackPoint::length>(length); }
+
+  inline double originalIndexByIndex(double index) const {
+    double frac;
+    const TTrackPoint &p0 = floorPoint(index, &frac);
+    const TTrackPoint &p1 = ceilPoint(index);
+    return interpolationLinear(p0.originalIndex, p1.originalIndex, frac);
+  }
+  inline double timeByIndex(double index) const {
+    double frac;
+    const TTrackPoint &p0 = floorPoint(index, &frac);
+    const TTrackPoint &p1 = ceilPoint(index);
+    return interpolationLinear(p0.time, p1.time, frac);
+  }
+  inline double lengthByIndex(double index) const {
+    double frac;
+    const TTrackPoint &p0 = floorPoint(index, &frac);
+    const TTrackPoint &p1 = ceilPoint(index);
+    return interpolationLinear(p0.length, p1.length, frac);
+  }
+
+  TTrackPoint calcPoint(double index) const;
+  TPointD calcTangent(double index, double distance = 0.1) const;
+  double rootIndexByIndex(double index) const;
+  TTrackPoint calcRootPoint(double index) const;
+
+  inline TTrackPoint interpolateLinear(double index) const {
+    double frac;
+    const TTrackPoint &p0 = floorPoint(index, &frac);
+    const TTrackPoint &p1 = ceilPoint(index);
+    return interpolationLinear(p0, p1, frac);
+  }
+
+  template<typename T>
+  static inline T interpolationLinear(const T &p0, const T &p1, double l)
+    { return p0*(1.0 - l) + p1*l; }
+
+  template<typename T>
+  static T interpolationSpline(const T &p0, const T &p1, const T &t0, const T &t1, double l) {
+    double ll = l*l;
+    double lll = ll*l;
+    return p0*( 2.0*lll - 3.0*ll + 1.0)
+         + p1*(-2.0*lll + 3.0*ll      )
+         + t0*(     lll - 2.0*ll + l  )
+         + t1*(     lll - 1.0*ll      );
+  }
+
+  static inline TTrackPoint interpolationLinear(const TTrackPoint &p0, const TTrackPoint &p1, double l) {
+    if (l <= TConsts::epsilon) return p0;
+    if (l >= 1.0 - TConsts::epsilon) return p1;
+    return TTrackPoint(
+      interpolationLinear(p0.position      , p1.position      , l),
+      interpolationLinear(p0.pressure      , p1.pressure      , l),
+      interpolationLinear(p0.tilt          , p1.tilt          , l),
+      interpolationLinear(p0.worldPosition , p1.worldPosition , l),
+      interpolationLinear(p0.screenPosition, p1.screenPosition, l),
+      interpolationLinear(p0.originalIndex , p1.originalIndex , l),
+      interpolationLinear(p0.time          , p1.time          , l),
+      interpolationLinear(p0.length        , p1.length        , l) );
```

Comment:
@blackwarthog 
The magnetism < 1.0 issue occurs because the `leftButtonUp()` isn't being called for Vector (and Toonz Rasters) levels.  This is because the TTrackPoint `final` variable, set during onRelease() is being lost along the way. 

I've tracked it to this point in the code when a new TTrackPoint is created and returned, it is ignoring the fact that p0 or p1 may have `final = true`.  

I can correct the issue by doing this, though not 100% sure this is the right fix nor the right place to deal with that variable:

>    return TTrackPoint(
>      interpolationLinear(p0.position      , p1.position      , l),
> ...
>      interpolationLinear(p0.length        , p1.length        , l),
>      **p0.final || p1.final** );